README
== Zena
CMS based on Ruby on Rails, (c) Gaspard Bucher, teti.ch 2007-2009, licenced under MIT

This version of zena is *TOTALLY UNUSABLE* for the moment. It is the result of the slow migration from Rails 1.2.3 to Rails 2.x. You will need the latest rails edge with some patches which you can find here: http://github.com/gaspard/rails/tree/master.

== Creating the database
You need to set collation and charset to utf8 when creating the database. Example :
# create database zena DEFAULT CHARSET=utf8 COLLATE=utf8_general_ci;

== Development/testing
1. Create the database using a mysql client
   create database zena_dev DEFAULT CHARSET=utf8 COLLATE=utf8_general_ci;

2. Rename 'config/database_example.yml' to 'config/database.yml' and update it's content

3. Migrate the database and add initialize db (from within zena directory)
   rake zena:migrate

4. Create a first site or load fixtures
   rake zena:mksite HOST=localhost PASSWORD=secret LANG=en

5. Start server
   script/server

6. Connect to site 'http://localhost:3000'

7. Login with 'admin' and 'secret'

== Production
You should use capistrano to ease deployment. See 'config/delploy.rb'.

Manual deployment is done with the following commands:
1. Create the database with (using a mysql client)
   create database zena DEFAULT CHARSET=utf8 COLLATE=utf8_general_ci;

2. Create log directory
   mkdir log

3. Rename 'config/database_example.yml' to 'config/database.yml' and update it's content
   (change user, password if needed)

4. Migrate the database
   rake zena:migrate RAILS_ENV=production

4. Create a site
   rake zena:mksite HOST=example.com PASSWORD=secret LANG=en RAILS_ENV=production

5. Login user is 'admin', password is the one set on the line above.

== Create a new site
# rake zena:mksite HOST=example.com PASSWORD=secret RAILS_ENV=production

== Start
# ruby lib/upload_progress_server.rb start
# script/server -S config/mongrel_upload_progress.conf

== Stop
# ^C  (interrupt)
# ruby lib/upload_progress_server.rb stop


== Login
You can now login with 'admin' and the password you used to make the site.

== Dependencies

=== 1. Original in debian etch (might need to replace jpeg by libjpeg62)
  # aptitude install build-essential apache2 mysql-server libmagick9-dev gs-gpl libssl-dev gettext libgettext-ruby libreadline5 libreadline5-dev zlib1g-dev libncurses5 libncurses5-dev unzip liburi-perl libjpeg-dev subversion ssh sudo awstats

If you want to enable LateX (PDF generation)
tetex-bin tetex-extra latex-ucs

For math mode (inline formulas), you will need
latex dviutils gs-gpl imagemagick   # use 'dvips' if dviutils is not found. Use 'tetex' if 'latex' not found.

For xsl-fo pdf generation
xsltproc fop # (contrib sources needed on debian for 'fop')
             # use 'libxslt' if the package 'xsltproc' does not exist for your platform.

# ocaml-base-nox can be replaced by 'ocaml'
=== 2. To be installed by hand
ruby1.8.6
    # wget ftp://ftp.ruby-lang.org/pub/ruby/1.8/ruby-1.8.6.tar.gz
    # tar xzf ruby-1.8.6.tar.gz
    # cd ruby-1.8.6
    # ./configure --with-openssl
    # make && make install
rubygems
    # wget http://rubyforge.org/frs/download.php/45905/rubygems-1.3.1.tgz
    # tar xzf rubygems-1.3.1.tgz
    # cd rubygems-1.3.1
    # ruby setup.rb (if ruby is not found, log out and log back in)

ImageMagick (on Linux)
    # wget ftp://ftp.imagemagick.org/pub/ImageMagick/ImageMagick.tar.gz
    # tar xzf ImageMagick.tar.gz
    # cd ImageMagick-*
    # ./configure
    # make && make install

ImageMagick (on mac os X using macports)
    # sudo port install tiff -macosx imagemagick +q8 +gs +wmf

=== 3. install Gems

You can remove "--no-ri --no-rdoc" if you want the documentation
    # gem install querybuilder rake hoe gettext mongrel mongrel_cluster rmagick tzinfo syntax mongrel_upload_progress uuidtools daemons json capistrano yamltest ruby-debug --no-ri --no-rdoc

Versions:

 querybuilder >= 0.5.5
 uuidtools >= 2.x

=== Advised tools
monit (debian package to monitor your mongrel processes)


== Very experimental webDAV
Should not be used for production.

add this to nodes_controller macros:
  act_as_railsdav
  include Zena::WebDav

add this to 'config/environment.rb'
  require File.join(File.dirname(__FILE__), '../lib/webdav_adapter')

install railsdav
  # script/plugin install http://svn.liverail.net/svn/plugins/railsdav

install dependencies
  # sudo gem install mime-types, unicode

History.txt
In this history, elements listed with a '~' still need to be documented.

== 1.3.0
  * Add Bundler Gemfile
  * works on MySQL 5.7
  * Update rmagick
  * Remove zena gem_config and gem_setup

== 1.2.9

* Minor changes
  * Fixed bug related to "data-t" attribute missing in <body> tag (#80).

== 1.2.8 2014-08-17

* Major changes
  * Added "update_attributes_without_clone" and "save_without_clone" to update attributes
    without versioning (used to sync attributes with external app). This does not change
    timestamps nor version or node author.
  * Added login retry wait delay on failures.
  * Added support for 'in home' or 'from home' in sqliss.
  * Added support for any scope with 'in sub_nodes' (uses fullpath field).
  ~ Added forced skin setting for sites (overwrites all but ACL skin settings).
  ~ Added support for profile users.
  * Added support to edit auth_user (linked user) through node's pseudo attribute 'auth' (admin only) or
    through 'auth_user'.
  ~ Created 'manager' user status: can edit users (only through 'auth').
  ~ Created 'currency' brick to fetch openexchangerates.org exchange rates.
  * Do not render an older Ajax reply if a newer one has already been received. This fixes the
    live query confusion.
  * Fixed loose requirements of httparty and rubyzip preventing installation.
  * Added 'fsize' method in view to display human readable octet size.
  ~ Added 'pasteTable' option to disable (by setting false) multi cell paste in grid.
  ~ Using contenteditable in grid instead of input form. Turn off with contenteditable option.
  * Fixed grid bug when leaving page while in input field.
  ~ Implemented 'fulltext match' or 'sphinx match' to use fulltext search in node query.
  ~ Added 'group_id_to_name' to visitor (enables group name display from node auth settings).
  ~ Added 'site_readonly' attribute to site with 'site.readonly?' rubyless. Useful for locking site during updates or migrations.

* Minor changes
  * Fixed gemspec to not include TextMate helper and selenium plugin.
  * Fixed xml generated file caching.
  ~ Added NODE_ID in 'redir' param on node creation (replaced by actual id).
  * Added fetch_html rubyless method to views (does an HTTP GET and returns body).
  * Fixed a bug in site alias zafu compilation (wrong home reference).
  * Fixed anchor in zazen url "":22#foo should link to anchor #foo in node '22'.
  * Prop eval returning nil for title properly replaces title with class name.
  * Type ctrl+return to enter a newline inside a grid.
  * Checking for defined accessors on links before passing to relation proxy (this enables custom code
    on relation modification without passing through RelationProxy).
  * Refuses to delete user node unless visitor is a manager.
  * Marks user as deleted and remove profile if node is destroyed.
  ~ Enable 'parent_id', 'project_id', 'section_id' Rubyless methods.
  * Fixed a critical bug related to changing a virtual class' superclass.
  * Do not trigger Ajax 'loading' unless it takes more then 300 ms (avoid blink).
  * Fixed tlabel on [input] when there is a default label (should overwride default).
  * Fixed "<p do='input'..." so that the "p" tag is preserved.
  * Using shift+enter to insert carriage return in grid.
  * Improved jobs display.
  * Fixed a cache handling bug where a page could exist in the filesystem without having a cache entry
    in the database.
  * Fixed building templates on the fly with fs_skin on JS call.
  * Fixed rebuild_fullpath and added link in sites on web interface.
  * Removing bricks.yml and other initial config files from 'rake zena:assets'.
  * Request for 'login' page while logged in redirects to home page instead of clearing session (fixes iPhone login issues).

== 1.2.6, 1.2.7 2013-08-16

* Major changes
  * Support for multiple domains (site alias) for a single site.
  * Added support for wkhtmltopdf (Thanks Jean-Baptiste).

* Minor changes
  ~ Added 'cmd' option for pdf brick.
  ~ Added 'local_ips' option to bricks.yml (used when pdf engine cannot render non-public assets).
  * Fixed dependency on newer versions of rails.

== 1.2.5 2013-07-15

* Major changes
 * To support video streaming: redirect on first document cache. WARNING: This
   implies changes to vhost files !
 * WARN: Properties in list context do not resolve with first element anymore. "first" prefix needed.
 * Enable VirtualClass on native classes !
 ~ Added 'activity' brick which records visitor's last "seen_at" date.
 ~ Support for site alias (other domain pointing to same data) "rake mkalias".

* Minor changes
 * Fixed encode_params in Ajax requests.
 * Fixed cached path in vhost file.
 ~ Added "uuid" method (generates a unique id).
 ~ Added support for JSON String in [string_hash]
 * Added 'login_info' to Node.
 * Added 'group_names' to User class.
 * Fixed [headers] when 'Status' is set (change to redirect on 300).

== 1.2.4 2013-06-13

* Major changes
 ~ Added class filtering to Acl in 'create' action.
 * Uploading html files does not transform them into zafu (use .zafu ext for this)
 * Added file upload support from zena remote API.
 ~ Added 'content_type' regexp to force virtual class on Document creation.
 * Extended "create group" in vclass to be "allow group" (edit).
 * Added support for [versions_list] to display list of versions and status.

* Minor changes
 * Support for raw html in zazen with '<notextile>' or '<html>' tag. Must be allowed with notextile='true' on [zazen] tag.
 * Fixed preview of content with ACL (considering the POST on /zafu as a 'read').
 * Added support for "class not like Image" or "class <> Image" to sqliss.
 ~ Added url to clear cache with /sites/clear_cache (admin only).
 * encode_params now supports arrays.
 * Add 'id' to date input.
 * Fixed 404 error.
 * Toggle takes dynamic parameters for "js" and "arity".
 * Fixed multiple toggles side-by-side.
 * Fixed nested blocks and class scoping.
 * Added support for "onUpdate" in [input] with date type.

db/schema.rb
# This file is auto-generated from the current state of the database. Instead of editing this file, 
# please use the migrations feature of Active Record to incrementally modify your database, and
# then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your database schema. If you need
# to create the application database on another system, you should be using db:schema:load, not running
# all the migrations from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended to check this file into your version control system.

ActiveRecord::Schema.define(:version => 20100115134729) do

  create_table "access_hits", :force => true do |t|
    t.integer  "site_id"
    t.integer  "node_id"
    t.string   "remote_host",      :limit => 50
    t.datetime "request_time"
    t.integer  "request_duration"
    t.integer  "status"
    t.integer  "bytes_sent"
    t.string   "request_method",   :limit => 6
    t.string   "request_uri"
    t.string   "referer"
    t.string   "agent"
    t.string   "request_line"
    t.string   "mode",             :limit => 30
    t.string   "format",           :limit => 10
    t.string   "lang",             :limit => 6
  end

  create_table "cached_pages", :force => true do |t|
    t.text     "path"
    t.datetime "expire_after"
    t.datetime "created_at"
    t.integer  "node_id"
    t.integer  "site_id"
  end

  create_table "cached_pages_nodes", :id => false, :force => true do |t|
    t.integer "cached_page_id"
    t.integer "node_id"
  end

  create_table "caches", :force => true do |t|
    t.datetime "updated_at"
    t.integer  "visitor_id"
    t.string   "visitor_groups", :limit => 200
    t.string   "kpath",          :limit => 200
    t.integer  "context"
    t.text     "content"
    t.integer  "site_id"
  end

  create_table "comments", :force => true do |t|
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "status",                       :default => 70, :null => false
    t.integer  "discussion_id"
    t.integer  "reply_to"
    t.integer  "user_id"
    t.string   "title",         :limit => 250, :default => "", :null => false
    t.text     "text",                                         :null => false
    t.string   "author_name",   :limit => 300
    t.integer  "site_id"
    t.string   "ip",            :limit => 200
  end

  create_table "contact_contents", :force => true do |t|
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "version_id"
    t.string   "first_name", :limit => 60,  :default => "", :null => false
    t.string   "name",       :limit => 60,  :default => "", :null => false
    t.text     "address"
    t.string   "zip",        :limit => 20,  :default => "", :null => false
    t.string   "city",       :limit => 60,  :default => "", :null => false
    t.string   "telephone",  :limit => 60,  :default => "", :null => false
    t.string   "mobile",     :limit => 60,  :default => "", :null => false
    t.string   "email",      :limit => 60,  :default => "", :null => false
    t.date     "birthday"
    t.integer  "site_id"
    t.string   "country",    :limit => 100
  end

  create_table "data_entries", :force => true do |t|
    t.integer  "site_id"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "user_id"
    t.datetime "date"
    t.text     "text"
    t.decimal  "value_a",    :precision => 24, :scale => 8
    t.integer  "node_a_id"
    t.integer  "node_b_id"
    t.integer  "node_c_id"
    t.integer  "node_d_id"
    t.decimal  "value_b",    :precision => 24, :scale => 8
  end

  create_table "discussions", :force => true do |t|
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "node_id"
    t.boolean  "inside",                   :default => false
    t.boolean  "open",                     :default => true
    t.string   "lang",       :limit => 10, :default => "",    :null => false
    t.integer  "site_id"
  end

  create_table "document_contents", :force => true do |t|
    t.string  "type",         :limit => 32
    t.integer "version_id"
    t.string  "name",         :limit => 200, :default => "", :null => false
    t.string  "content_type", :limit => 40
    t.string  "ext",          :limit => 20
    t.integer "size"
    t.integer "width"
    t.integer "height"
    t.integer "site_id"
    t.text    "exif_json"
  end

  create_table "dyn_attributes", :force => true do |t|
    t.integer "owner_id"
    t.string  "owner_table"
    t.string  "key"
    t.text    "value"
  end

  add_index "dyn_attributes", ["owner_id"], :name => "index_dyn_attributes_on_owner_id"
  add_index "dyn_attributes", ["owner_table"], :name => "index_dyn_attributes_on_owner_table"

  create_table "groups", :force => true do |t|
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "name",       :limit => 20, :default => "", :null => false
    t.integer  "site_id"
  end

  create_table "groups_users", :id => false, :force => true do |t|
    t.integer "group_id"
    t.integer "user_id"
  end

  create_table "iformats", :force => true do |t|
    t.string   "name",       :limit => 40
    t.integer  "site_id"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "size"
    t.integer  "gravity"
    t.integer  "width"
    t.integer  "height"
    t.string   "popup",      :limit => 120
  end

  create_table "links", :force => true do |t|
    t.integer  "source_id"
    t.integer  "target_id"
    t.integer  "relation_id"
    t.integer  "status"
    t.string   "comment",     :limit => 60
    t.datetime "date"
  end

  create_table "nodes", :force => true do |t|
    t.string   "name",         :limit => 200
    t.string   "type",         :limit => 32
    t.integer  "vclass_id"
    t.string   "kpath",        :limit => 16
    t.datetime "event_at"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "user_id"
    t.integer  "parent_id"
    t.integer  "project_id"
    t.integer  "section_id"
    t.string   "skin"
    t.integer  "inherit"
    t.integer  "rgroup_id"
    t.integer  "wgroup_id"
    t.integer  "dgroup_id"
    t.datetime "publish_from"
    t.datetime "log_at"
    t.string   "ref_lang",     :limit => 10,  :default => "",    :null => false
    t.string   "alias",        :limit => 400
    t.text     "fullpath"
    t.boolean  "custom_base",                 :default => false
    t.text     "basepath"
    t.integer  "site_id"
    t.integer  "zip"
    t.float    "position",                    :default => 0.0
    t.integer  "custom_a"
    t.integer  "custom_b"
    t.text     "vhash"
  end

  create_table "relations", :force => true do |t|
    t.string  "source_role",   :limit => 32
    t.string  "source_kpath",  :limit => 16
    t.boolean "source_unique"
    t.string  "source_icon",   :limit => 200
    t.string  "target_role",   :limit => 32
    t.string  "target_kpath",  :limit => 16
    t.boolean "target_unique"
    t.string  "target_icon",   :limit => 200
    t.integer "site_id",                      :null => false
  end

  create_table "sessions", :force => true do |t|
    t.string   "session_id", :null => false
    t.text     "data"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "sessions", ["session_id"], :name => "index_sessions_on_session_id"
  add_index "sessions", ["updated_at"], :name => "index_sessions_on_updated_at"

  create_table "site_attributes", :force => true do |t|
    t.integer "owner_id"
    t.string  "key"
    t.text    "value"
  end

  add_index "site_attributes", ["owner_id"], :name => "index_site_attributes_on_owner_id"

  create_table "sites", :force => true do |t|
    t.string   "host"
    t.integer  "root_id"
    t.integer  "su_id"
    t.integer  "anon_id"
    t.integer  "public_group_id"
    t.integer  "site_group_id"
    t.string   "name"
    t.boolean  "authentication"
    t.string   "languages"
    t.string   "default_lang"
    t.boolean  "http_auth"
    t.boolean  "auto_publish"
    t.integer  "redit_time"
    t.datetime "formats_updated_at"
  end

  create_table "template_contents", :force => true do |t|
    t.integer "site_id"
    t.integer "node_id"
    t.string  "skin_name"
    t.string  "format"
    t.string  "tkpath"
    t.string  "klass"
    t.string  "mode"
  end

  create_table "users", :force => true do |t|
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "login",               :limit => 20
    t.string   "crypted_password",    :limit => 40
    t.string   "first_name",          :limit => 60
    t.string   "name",                :limit => 60
    t.string   "email",               :limit => 60
    t.string   "time_zone"
    t.integer  "site_id"
    t.integer  "status"
    t.integer  "contact_id"
    t.string   "lang",                :limit => 10, :default => "", :null => false
    t.string   "persistence_token"
    t.string   "password_salt"
    t.string   "single_access_token"
  end

  create_table "versions", :force => true do |t|
    t.string   "type",         :limit => 32
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "node_id"
    t.integer  "user_id"
    t.string   "lang",         :limit => 10, :default => "", :null => false
    t.datetime "publish_from"
    t.text     "comment"
    t.text     "title"
    t.text     "summary"
    t.text     "text"
    t.integer  "status",                     :default => 70, :null => false
    t.integer  "number",                     :default => 1,  :null => false
    t.integer  "content_id"
    t.integer  "site_id"
  end

  if Zena::Db.adapter == 'mysql'
    execute "ALTER TABLE versions ENGINE = MyISAM"
    execute "CREATE FULLTEXT INDEX index_versions_on_title_and_text_and_summary ON versions (title,text,summary)"
  end

  create_table "virtual_classes", :force => true do |t|
    t.string  "name"
    t.string  "kpath",                  :limit => 16
    t.string  "real_class",             :limit => 16
    t.string  "icon",                   :limit => 200
    t.integer "create_group_id"
    t.integer "site_id",                               :null => false
    t.boolean "auto_create_discussion"
    t.text    "dyn_keys"
  end

  create_table "zips", :id => false, :force => true do |t|
    t.integer "site_id"
    t.integer "zip"
  end

end
